### PR TITLE
GH-2640/transformer forward

### PR DESCRIPTION
**Pull request by helpmefindaname**
Another PR as part of https://github.com/flairNLP/flair/issues/2640
resructures the TransformerEmbeddings to be exportable as Onnx model.
This PR also reworks the subtoken2token mapping, as the huggingface tokenizers provide "word_ids" which can be used for this (see their new Book for details)

I consider the following should be part of the PR:
- [x] capabilities of exporting functional Onnx models.
- [x] consider refactoring context for flert 
- [x] extensive testing of all pooling variations and possibilities
- [x] performance testing of the new TransformerEmbeddings compared to the old once
- [x] implementation of OnnxTransformerEmbeddings as drop in replacement for TransformerEmbeddings  
- [x] examples for the standard optimizations in the OnnxRuntime (fusion & dynamic quantisation)
- [x] performance testing of the Onnx models
- [x] refactoring (don't have everything in embeddings.base.py)

### CODE REVIEW

**Comment by helpmefindaname** [comment]
results from the performance testing:
grafik
grafik

I tested using the FLERT model on the first 200 (gpu) or 64 (cpu) sentences of the conll03 test-set.
with a batch-size of 1 (gpu) or 32 (cpu).

the speed is measured as "seconds per sentences" (lower is better) and the results show that there are slight but consistent improvements in all modes when comparing the master with this refactored version. (however cpu only is less consistent than others).
as quantisation was always possible, it was easy to compare new vs old.
for jit only the refactoring provides possibilities and that also only with some hacking:

first the embedding needs to be covered by a Jitable Wrapper:
```
class JitWrapper(TokenEmbeddings):
    def __init__(self, emb: TransformerWordEmbeddings):
        super().__init__()
        self.emb = emb

    @property
    def embedding_length(self) -> int:
        return self.emb.embedding_length

    def forward(
        self,
        input_ids: torch.Tensor,
        attention_mask: torch.Tensor,
        overflow_to_sample_mapping: torch.Tensor,
        word_ids: torch.Tensor,
    ):
        return self.emb.forward(
            input_ids=input_ids,
            attention_mask=attention_mask,
            overflow_to_sample_mapping=overflow_to_sample_mapping,
            word_ids=word_ids,
        )["token_embeddings"]

    def get_names(self) -> List[str]:
        return self.emb.get_names()

    def _add_embeddings_internal(self, sentences: List[Sentence]):
        tensors = self.emb._prepare_tensors(sentences)

        with torch.no_grad():
            embeddings = self.forward(**tensors)
            self.emb._extract_token_embeddings(embeddings, sentences)
```
the important part is, that the `forward` method of the JitWrapper, has a similar signature to the `forward` of the `TransformerEmbeddings`, but removes all parameters that are set to None (for example it discards `langs`)


and in the model the jit functionallity needs to be called via `jit.trace`:
```
tensors = model.embeddings._prepare_tensors(sentences[:2])
model.embeddings = JitWrapper(model.embeddings)
params = []
for param in inspect.signature(model.embeddings.forward).parameters.keys():
    params.append(tensors[param])

model.embeddings.forward = torch.jit.trace(model.embeddings, params, strict=False).forward
```

**Comment by alanakbik** [comment]
@helpmefindaname this is great, thanks for sharing these findings. So "main" in the measurements is the current master/main branch, correct? For CUDA, does this mean that you see improvements from 0.054 seconds per sentence to 0.044 seconds per sentence with the refactoring? This would be almost 20% faster, which would be awesome!

**Comment by helpmefindaname** [comment]
> @helpmefindaname this is great, thanks for sharing these findings. So "main" in the measurements is the current master/main branch, correct? For CUDA, does this mean that you see improvements from 0.054 seconds per sentence to 0.044 seconds per sentence with the refactoring? This would be almost 20% faster, which would be awesome!

I think in this case it is more of a absolute improvement than a relative one. I suppose those 0.01s are mainly due to the rework of the subtoken matching. There was a lot of string matching which now is replaced by the word_ids created by the transformer tokenizers

**Comment by helpmefindaname** [comment]
@alanakbik I noticed that I didn't answer your question, yes you are correct about the "main". Btw this branch is now ready for review, would be awesome if we manage to get it merged soon, feel free to approach me on slack if there are any questions

**Comment by helpmefindaname** [comment]
as reported in https://github.com/flairNLP/flair/issues/2528 there is a case where using jit tracing via aws neutron "is it's performing almost 4x faster than a t4 GPU with 70% lower cost" 😄

**Comment by alanakbik** [comment]
@helpmefindaname my mistake, getting the error now. Checking.

**Comment by helpmefindaname** [comment]
@alanakbik as I got a notification that you did rerun the tests: I assume there is some incompatibility with the new torch `1.12.0` as it did work with `1.11.0` before. I will further investigate this

**Comment by helpmefindaname** [comment]
That is easy to explain:

torch.jit doesn't translate the python code to a c++ module, but traces the code by running it once and store which tensor operations were run to generate a computation graph. In that graph we will have plenty of constants, for example the line:
`hidden_states.permute((0, 3, 1, 2)), 0, 1)` makes use of 6 constants. As jit is tracing and not interpreting the python code, it doesn't know that those are actually constants. It could be that we would have code like the following:
```
def fn(hidden_states, a, b, c, d, e, f):
    return hidden_states.permute((a, b, c, d)), e, f)
```
Where a,b,c,d,e,f would be integer parameters. In that case jit would see exactly the same, even if it was intended, that those parameters can change over calls. To prevent us from unintentionally converting such parameters to constants in the graph, we get this warning.

I suppose we get this warnings due to parameters like `stride`, `allow_long_sequences`, `hidden_dimension`, ...

**Comment by helpmefindaname** [comment]
Okay, after long debugging I realized, that the system error was basically only the warning being cached as error by the pytest configuration. Hence the fix was very simple.
However a learning that I got:
the warning is only shown in pytorch 1.12.0 and it is due to the line:
` scores = scores.masked_fill(mask, torch.tensor(-float("inf")))` in the transformers package. Here a tensor is explicitly created with a constant value, which leads to the warning.

File: resources/docs/TUTORIAL_13_TRANSFORMERS_PRODUCTION.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
+# Speed up prediction speed of Transformer Embeddings
```
Comment (alanakbik) [inline resources/docs/TUTORIAL_13_TRANSFORMERS_PRODUCTION.md:1]: Awesome, thanks a lot for writing a tutorial for the feature!

**Review by alanakbik — APPROVED** [review]
Thanks a lot for adding this feature - this will surely be useful to many people!

**Comment by heukirne** [comment]
Hello @helpmefindaname. 
First of all, thank you so much for this PR. It will be useful for several people!

I'm trying to follow your [tutorial](https://github.com/flairNLP/flair/blob/master/resources/docs/TUTORIAL_13_TRANSFORMERS_PRODUCTION.md) but I facing this issue:
`_forward_unimplemented() got an unexpected keyword argument 'input_ids'`
at
`script_module = torch.jit.trace(wrapper, parameter_list)`

I have a model trained with TransformerWordEmbeddings Flair 0.11
Should I train it again with last merged version Flair 0.11.3 (master)?

Cheers!

**Comment by heukirne** [comment]
@helpmefindaname , training and tracing with Flair 0.11.3 (master) it works!
